stats/cloud/api.go

```diff
@@ -89,40 +93,66 @@ func (c *Client) CreateTestRun(testRun *TestRun) (*CreateTestRunResponse, error)
 }
 
 func (c *Client) PushMetric(referenceID string, noCompress bool, samples []*Sample) error {
+	start := time.Now()
 	url := fmt.Sprintf("%s/metrics/%s", c.baseURL, referenceID)
 
-	if noCompress {
-		req, err := c.NewRequest("POST", url, samples)
-		if err != nil {
-			return err
-		}
-		return c.Do(req, nil)
+	jsonStart := time.Now()
+	b, err := json.Marshal(&samples)
+	if err != nil {
+		return err
 	}
+	jsonTime := time.Since(jsonStart)
 
-	var buf bytes.Buffer
-	var unzippedSize int
-	if samples != nil {
-		b, err := json.Marshal(&samples)
-		if err != nil {
-			return err
-		}
-		unzippedSize = len(b)
-		g := gzip.NewWriter(&buf)
-		if _, err = g.Write(b); err != nil {
-			return err
-		}
-		if err = g.Close(); err != nil {
-			return err
-		}
-	}
-	req, err := http.NewRequest("POST", url, &buf)
+	// TODO: change the context, maybe to one with a timeout
+	req, err := http.NewRequestWithContext(context.Background(), "POST", url, nil)
 	if err != nil {
 		return err
 	}
-	req.Header.Set("Content-Encoding", "gzip")
-	req.Header.Set("x-payload-byte-count", strconv.Itoa(unzippedSize))
-	req.Header.Set("x-payload-sample-count", strconv.Itoa(len(samples)))
-	return c.Do(req, nil)
+
+	req.Header.Set("X-Payload-Sample-Count", strconv.Itoa(len(samples)))
+	var additionalFields logrus.Fields
+
+	if !noCompress {
+		buf := bytes.NewBuffer(nil) // use pool
+		unzippedSize := len(b)
+		//nolint:gomnd
+		buf.Grow(unzippedSize / 5) // probably much smaller
+		gzipStart := time.Now()
+		{
+			g := gzip.NewWriter(buf)
+			if _, err = g.Write(b); err != nil {
+				return err
+			}
+			if err = g.Close(); err != nil {
+				return err
+			}
+		}
+		gzipTime := time.Since(gzipStart)
+		req.Header.Set("Content-Encoding", "gzip")
+		req.Header.Set("X-Payload-Byte-Count", strconv.Itoa(unzippedSize))
+		req.Header.Set("Content-Length", strconv.Itoa(buf.Len()))
+
+		additionalFields = logrus.Fields{
+			"unzipped_size":  unzippedSize,
+			"gzip_t":         gzipTime,
+			"content_length": buf.Len(),
+		}
+
+		req.Body = ioutil.NopCloser(buf)
+	} else {
+		req.Header.Set("Content-Length", strconv.Itoa(len(b)))
+		req.Body = ioutil.NopCloser(bytes.NewReader(b))
```

Comment:
A minor nitpick, but if you set `GetBody()` here, you can add a check in `client.go` not to read the whole body in a new buffer again. It's probably going do be a substantial optimization if we have multi-megabyte requests.

Reply:
Hm ... I did think of that, and even started it, including [your other comment](https://github.com/loadimpact/k6/pull/1569#discussion_r460794124), and then ended with the current code ...

But you are right ... I did think there is an optimization to be had ... I will make the change in a bit

Reply:
I did it slightly differently in https://github.com/loadimpact/k6/pull/1569/commits/b57bec26e4a1d2c0ccf3bd0e054f30cf4b416679, but I think this is even better
1. `b` no longer keeps the uncompressed marshalled json 
2. If something added GetBody, they won't use the cloud.client implementaiton